core/slice/ptr.odin

```diff
@@ -73,25 +73,17 @@ ptr_rotate :: proc(left: int, mid: ^$T, right: int) {
 		left, mid, right := left, mid, right
 
 		// TODO(bill): Optimization with a buffer for smaller ranges
-		if left >= right {
-			for {
-				ptr_swap_non_overlapping(ptr_sub(mid, right), mid, right)
+		for left > 0 && right > 0 {
+			if left >= right {
+				ptr_swap_non_overlapping(ptr_sub(mid, right), mid, right * size_of(T))
 				mid = ptr_sub(mid, right)
 
 				left -= right
-				if left < right {
-					break
-				}
-			}
-		} else {
-			for {
-				ptr_swap_non_overlapping(ptr_sub(mid, left), mid, left)
+			} else {
+				ptr_swap_non_overlapping(ptr_sub(mid, left), mid, left * size_of(T))
```

Comment:
Why have you changed this logic?

Reply:
The logic needed to be changed because it's possible for the ordering of the left and right sizes to change multiple times before the algorithm terminates.
With the privious logic:
```
[1 2 3 4 5 6 7 8]
|      ^        |
iteration 1: left = 3, right = 5
swaps the first 3 with the following 3
[4 5 6 1 2 3 7 8]
       |     ^  |
iteration 2: left = 3, right = 2
range is not fully rotated, we still need to swap [1 2 3] with [7 8],
however the previous logic terminated here due to right becoming smaller than left.

```